src/lib.rs

```diff
@@ -0,0 +1,238 @@
+                    let k = BytesKey::from(actor.to_bytes()?);
+                    if !entry.contains_key(&k)? {
+                        return Ok(None);
+                    }
+                }
+
+                // Prepare messages to commit the atomic execution
+                let mut msgs = Vec::new();
+                entry.for_each(|addr, &method| {
+                    msgs.push(CrossMsg {
+                        msg: StorableMsg {
+                            to: IPCAddress::from_bytes(addr)?,
+                            method,
+                            params: exec_id.to_bytes().into(),
+                            ..Default::default()
+                        },
+                        wrapped: true,
+                    });
+                    Ok(())
+                })?;
+                Ok(Some(msgs))
+            })
+            .map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to update registry")
+            })
+        })?;
+
+        match msgs {
+            Some(msgs) => {
+                // Send the messages to commit the atomic execution
+                for msg in msgs {
+                    rt.send(
+                        st.ipc_gateway_address,
+                        ipc_gateway::Method::SendCross as MethodNum,
+                        RawBytes::serialize(msg)?,
+                        TokenAmount::zero(),
+                    )?;
+                }
+
+                // Remove the atomic execution entry
+                rt.transaction(|st: &mut State, rt| {
+                    st.rm_atomic_exec(rt.store(), *exec_id, actors.clone())
+                        .map_err(|e| {
+                            e.downcast_default(
+                                ExitCode::USR_ILLEGAL_STATE,
+                                "failed to remove atomic exec from registry",
+                            )
+                        })
+                })?;
+
+                Ok(true)
+            }
+            None => Ok(false),
+        }
+    }
+
+    /// Removes a pre-commitment from an actor to perform an atomic
+    /// execution. This method is to be invoked by a wrapped crossnet
+    /// message originating in one of the execution actors involved in
+    /// the atomic execution.
+    fn revoke<BS, RT>(rt: &mut RT, params: ApplyMsgParams) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        rt.validate_immediate_caller_is(std::iter::once(&st.ipc_gateway_address))?;
+
+        let ApplyMsgParams {
+            cross_msg:
+                CrossMsg {
+                    msg: StorableMsg { from, params, .. },
+                    ..
+                },
+        } = params;
+
+        let params: RevokeParams = cbor::deserialize_params(&params)?;
+        let actors = &params.actors;
+        let exec_id = &params.exec_id;
+
+        if !actors.contains(&from) {
```

Comment:
Here we check that we receive a pre-commitment from all actors, but how do we know that they reached the same output? And how to actors know what state to commit?

Reply:
The atomic execution ID `exec_id` must uniquely determine the output:

> ... The execution actor further computes the atomic execution output and
stores it in its state associated with the atomic execution ID.

Reply:
Right, but there is more than one actor involved in the execution, and each of them live in a different subnet, how do they reach the same `exec_id`? I guess that through the input parameters for the execution, right? But this is not obvious as this is not shared in any way with the coordinator (let me know if my point is not clear, I can try to follow-up or we can have a sync)

Reply:
Exactly, all of the execution actors are supposed to receive the same set of inputs when invoked for pre-commitment:

> The method deterministically computes a concise _atomic execution ID_ which must be unique for each distinct set of inputs (including actor IDs) and such that all the execution actors compute the same ID.

The key point is that the actors must deterministically produce the same `exec_id`. It is _the responsibility of the execution actors and the users_ to ensure that in order to complete the atomic execution successfully.